dali/operators/numba_function/numba_func_gpu.cc

```diff
@@ -115,6 +75,79 @@ NumbaFuncImpl<GPUBackend>::NumbaFuncImpl(const OpSpec &spec) : Base(spec) {
   }
 }
 
+
+template <>
+void NumbaFuncImpl<GPUBackend>::OutputsSetupFn(std::vector<OutputDesc> &output_desc,
+                                               int noutputs, int ninputs, int nsamples) {
+  out_shapes_.resize(noutputs);
+  for (int i = 0; i < noutputs; i++) {
+    out_shapes_[i].resize(nsamples, outs_ndim_[i]);
+    output_desc[i].type = static_cast<DALIDataType>(out_types_[i]);
+  }
+
+  input_shape_ptrs_.resize(nsamples * ninputs);
+  for (int in_id = 0; in_id < ninputs; in_id++) {
+    for (int i = 0; i < nsamples; i++) {
+      input_shape_ptrs_[nsamples * in_id + i] =
+        reinterpret_cast<uint64_t>(in_shapes_[in_id].tensor_shape_span(i).data());
+    }
+  }
+
+  output_shape_ptrs_.resize(nsamples * noutputs);
+  for (int out_id = 0; out_id < noutputs; out_id++) {
+    for (int i = 0; i < nsamples; i++) {
+      output_shape_ptrs_[nsamples * out_id + i] =
+        reinterpret_cast<uint64_t>(out_shapes_[out_id].tensor_shape_span(i).data());
+    }
+  }
+
+  ((void (*)(void*, const void*, int32_t, const void*, const void*, int32_t, int32_t))setup_fn_)(
+    output_shape_ptrs_.data(), outs_ndim_.data(), noutputs,
+    input_shape_ptrs_.data(), ins_ndim_.data(), ninputs, nsamples);
+
+  for (int out_id = 0; out_id < noutputs; out_id++) {
+    output_desc[out_id].shape = out_shapes_[out_id];
+    for (int i = 0; i < nsamples; i++) {
+      auto out_shape_span = output_desc[out_id].shape.tensor_shape_span(i);
+      for (int d = 0; d < outs_ndim_[out_id]; d++) {
+        DALI_ENFORCE(out_shape_span[d] >= 0, make_string(
+          "Shape of data should be non negative. ",
+          "After setup function shape for output number ",
+          out_id, " in sample ", i, " at dimension ", d, " is negative."));
+      }
+    }
+  }
+
+  out_strides_.clear();
+  out_strides_.reserve(noutputs * nsamples);
+  out_arrays_.clear();
+  out_arrays_.reserve(noutputs * nsamples);
+  for (int out_id = 0; out_id < noutputs; out_id++) {
+    for (int i = 0; i < nsamples; i++) {
+      out_strides_.push_back(kernels::GetStrides(out_shapes_[out_id][i]));
+      out_arrays_.push_back(
+        NumbaDevArray(out_shapes_[out_id].tensor_shape_span(i),
+                      make_span(out_strides_.back()), out_types_[out_id]));
+    }
+  }
+}
+
+
+template <>
+void NumbaFuncImpl<GPUBackend>::OutputsSetupNoFn(std::vector<OutputDesc> &output_desc,
+                                                 int noutputs, int ninputs, int nsamples) {
+  DALI_ENFORCE(noutputs == ninputs, make_string(
+      "Size of `out_types` should match size of `in_types` if the custom `setup_fn` function ",
+      "is not provided. Provided ", ninputs, " inputs and ", noutputs,
+      " outputs."));
```

Comment:
This message seems misplaced, since there are no "out_types" or "in_types" here.